tests/nlu/training_data/test_markdown.py

```diff
@@ -0,0 +1,267 @@
+from typing import Optional, Text, Dict, Any
+
+import pytest
+
+from rasa.nlu import load_data
+from rasa.nlu.extractors.crf_entity_extractor import CRFEntityExtractor
+from rasa.nlu.extractors.duckling_http_extractor import DucklingHTTPExtractor
+from rasa.nlu.extractors.mitie_entity_extractor import MitieEntityExtractor
+from rasa.nlu.extractors.spacy_entity_extractor import SpacyEntityExtractor
+from rasa.nlu.training_data.formats import RasaReader
+from rasa.nlu.training_data.formats import MarkdownReader, MarkdownWriter
+
+
+def test_markdown_entity_regex():
+    r = MarkdownReader()
+
+    md = """
+## intent:restaurant_search
+- i'm looking for a place to eat
+- i'm looking for a place in the [north](loc-direction) of town
+- show me [chines](cuisine:chinese) restaurants
+- show me [chines](22_ab-34*3.A:43er*+?df) restaurants
+    """
+
+    result = r.reads(md)
+
+    assert len(result.training_examples) == 4
+    first = result.training_examples[0]
+    assert first.data == {"intent": "restaurant_search"}
+    assert first.text == "i'm looking for a place to eat"
+
+    second = result.training_examples[1]
+    assert second.data == {
+        "intent": "restaurant_search",
+        "entities": [
+            {"start": 31, "end": 36, "value": "north", "entity": "loc-direction"}
+        ],
+    }
+    assert second.text == "i'm looking for a place in the north of town"
+
+    third = result.training_examples[2]
+    assert third.data == {
+        "intent": "restaurant_search",
+        "entities": [{"start": 8, "end": 14, "value": "chinese", "entity": "cuisine"}],
+    }
+    assert third.text == "show me chines restaurants"
+
+    fourth = result.training_examples[3]
+    assert fourth.data == {
+        "intent": "restaurant_search",
+        "entities": [
+            {"start": 8, "end": 14, "value": "43er*+?df", "entity": "22_ab-34*3.A"}
+        ],
+    }
+    assert fourth.text == "show me chines restaurants"
+
+
+def test_markdown_empty_section():
+    data = load_data("data/test/markdown_single_sections/empty_section.md")
+    assert data.regex_features == [{"name": "greet", "pattern": r"hey[^\s]*"}]
+
+    assert not data.entity_synonyms
+    assert len(data.lookup_tables) == 1
+    assert data.lookup_tables[0]["name"] == "chinese"
+    assert "Chinese" in data.lookup_tables[0]["elements"]
+    assert "Chines" in data.lookup_tables[0]["elements"]
+
+
+def test_markdown_not_existing_section():
+    with pytest.raises(ValueError):
+        load_data("data/test/markdown_single_sections/not_existing_section.md")
+
+
+def test_section_value_with_delimiter():
+    td_section_with_delimiter = load_data(
+        "data/test/markdown_single_sections/section_with_delimiter.md"
+    )
+    assert td_section_with_delimiter.entity_synonyms == {"10:00 am": "10:00"}
+
+
+def test_read_complex_entity_format():
+    r = MarkdownReader()
```

Comment:
let's add more tests for the complex format, e.g.
- failing schema validation (invalid json, missing keys)
- numbers, space, underscores, escaped quotes (?) in the dict values
- what happens if somebody does `"city": null`?
- messages with curly brackets

Reply:
we should also test the deprecation warning